RFS/src/main/java/com/rfs/common/DocumentReindexer.java

```diff
@@ -20,94 +22,101 @@
-  }
-
-  @EqualsAndHashCode(onlyExplicitlyIncluded = true)
-  public static class BulkDocSection {
-
-    @EqualsAndHashCode.Include
-    @Getter
-    private final String docId;
-    private final String bulkIndex;
-
-    public BulkDocSection(Document doc) {
-      this.docId = Uid.decodeId(doc.getBinaryValue("_id").bytes);
-      this.bulkIndex = createBulkIndex(docId, doc);
-    }
+    protected final OpenSearchClient client;
+    private final int maxDocsPerBulkRequest;
+    private final long maxBytesPerBulkRequest;
+    private final int maxConcurrentWorkItems;
+
+    public Mono<Void> reindex(String indexName, Flux<Document> documentStream, IDocumentReindexContext context) {
+        // Create scheduler for short-lived CPU bound tasks
+        var scheduler = Schedulers.newParallel("DocumentReader");
+        var bulkDocs = documentStream.publishOn(scheduler).map(BulkDocSection::new);
 
-    @SneakyThrows
-    private static String createBulkIndex(final String docId, final Document doc) {
-      // For a successful bulk ingestion, we cannot have any leading or trailing whitespace, and  must be on a single line.
-      String trimmedSource = doc.getBinaryValue("_source").utf8ToString().trim().replace("\n", "");
-      return "{\"index\":{\"_id\":\"" + docId + "\"}}" + "\n" + trimmedSource;
+        return this.reindexDocsInParallelBatches(bulkDocs, indexName, context)
+            .doOnSuccess(unused -> log.debug("All batches processed"))
+            .doOnError(e -> log.error("Error prevented all batches from being processed", e))
+            .doOnTerminate(scheduler::dispose);
     }
 
-    public static String convertToBulkRequestBody(Collection<BulkDocSection> bulkSections) {
-      StringBuilder builder = new StringBuilder();
-      for (var section : bulkSections) {
-        var indexCommand = section.asBulkIndex();
-        builder.append(indexCommand).append("\n");
-      }
-      return builder.toString();
+    Mono<Void> reindexDocsInParallelBatches(Flux<BulkDocSection> docs, String indexName, IDocumentReindexContext context) {
+        // Create elastic scheduler for long-lived i/o bound tasks
+        var scheduler = Schedulers.newBoundedElastic(maxConcurrentWorkItems, Integer.MAX_VALUE, "DocumentBatchReindexer");
+        var bulkDocsBatches = batchDocsBySizeOrCount(docs);
+        var maxGroupsToPrefetch = 1;
+
+        return bulkDocsBatches
+            .parallel(maxConcurrentWorkItems, maxConcurrentWorkItems /* ??? how is the related to the second arg on the next line ??? */)
+            .runOn(scheduler, maxGroupsToPrefetch)
+            .concatMapDelayError(docsGroup -> sendBulkRequest(UUID.randomUUID(), docsGroup, indexName, context))
+            .doOnTerminate(scheduler::dispose)
+            .then();
+        }
+
+    Mono<Void> sendBulkRequest(UUID batchId, List<BulkDocSection> docsBatch, String indexName, IDocumentReindexContext context) {
```

Comment:
Added in `batchId` to correlate the start message with the success/failure messages